### PR TITLE
[SPARK-48834][SQL] Disable variant input/output to python scalar UDFs, UDTFs, UDAFs during query compilation

### PR DESCRIPTION
**Pull request by richardc-db**
### What changes were proposed in this pull request?

Throws an exception if a variant is the input/output type to/from python UDF, UDAF, UDTF

### Why are the changes needed?

currently, variant input/output types to scalar UDFs will fail during execution or return a `net.razorvine.pickle.objects.ClassDictConstructor` to the user python code. For a better UX, we should fail during query compilation for failures, and block returning `ClassDictConstructor` to user code as we one day want to actually return `VariantVal`s to the user code. 

### Does this PR introduce _any_ user-facing change?

yes - attempting to use variants in python UDFs will now throw an exception rather than returning a `ClassDictConstructor` as before. However, we want to make this change now as we one day want to be able to return `VariantVal`s to the user code and do not want users relying on this current behavior

### How was this patch tested?

added UTs

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Review by allisonwang-db — COMMENTED** [review]
How about Python UDF/UDTFs?

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -954,6 +954,16 @@
           "The input of <functionName> can't be <dataType> type data."
         ]
       },
+      "UNSUPPORTED_UDF_INPUT_TYPE" : {
+        "message" : [
+          "UDFs do not support <dataType> type input data."
```
Comment (allisonwang-db) [inline common/utils/src/main/resources/error/error-conditions.json:959]:
```suggestion
          "UDFs do not support '<dataType>' as an input data type."
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 2]
```diff
@@ -954,6 +954,16 @@
...
+        ]
+      },
+      "UNSUPPORTED_UDF_OUTPUT_TYPE" : {
+        "message" : [
+          "UDFs do not support <dataType> type output data."
```
Comment (allisonwang-db) [inline common/utils/src/main/resources/error/error-conditions.json:964]:
```suggestion
          "UDFs do not support '<dataType>' as an output data type."
```

File: python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py [hunk 3]
```diff
@@ -748,6 +750,31 @@ def check_vectorized_udf_return_scalar(self):
...
+        from pyspark.sql.functions import col
+
+        scalar_f = pandas_udf(lambda u: str(u), StringType())
+        iter_f = pandas_udf(
+            lambda it: map(lambda u: str(u), it), StringType(), PandasUDFType.SCALAR_ITER
+        )
+        expectedErrorStr = 'UDFs do not support "VARIANT" type input data'
+        for f in [scalar_f, iter_f]:
+            with self.assertRaisesRegex(AnalysisException, expectedErrorStr):
+                df.select(f(col("v"))).collect()
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py:764]: Can we check the error class instead of the error string here?

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -194,16 +194,7 @@ def fromDDL(cls, ddl: str) -> "DataType":
...
-        from pyspark.sql.functions import udf
-
-        # Intentionally uses SparkSession so one implementation can be shared with/without
-        # Spark Connect.
-        schema = (
-            SparkSession.active().range(0).select(udf(lambda x: x, returnType=ddl)("id")).schema
-        )
-        assert len(schema) == 1
-        return schema[0].dataType
+        return _parse_datatype_string(ddl)
```
Comment (allisonwang-db) [inline python/pyspark/sql/types.py:197]: Why is this change needed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala [hunk 5]
```diff
@@ -63,6 +63,32 @@ trait PythonFuncExpression extends NonSQLExpression with UserDefinedExpression {
...
+      exprReturningVariant match {
+        case Some(e) => TypeCheckResult.DataTypeMismatch(
+          errorSubClass = "UNSUPPORTED_UDF_INPUT_TYPE",
+          messageParameters = Map("dataType" -> s"\"${e.dataType.sql}\""))
+        case None => TypeCheckResult.TypeCheckSuccess
+      }
+    }
+  }
+
+  def typeContainsVariant(dt: DataType): Boolean = dt match {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:84]: Do we have a helper function for this? This might be needed to check for other expressions.

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py (hunk 3 above)
Comment (richardc-db) [reply python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py:764]: done!

File: python/pyspark/sql/types.py (hunk 4 above)
Comment (richardc-db) [reply python/pyspark/sql/types.py:197]: If we want to do something like `fromDDL("v variant")`, `fromDDL` actually calls a `udf` just to get the output schema. However, this PR disables variant output UDFs during planning, so we have to get the data type another way

I've discussed a bit with @HyukjinKwon offline last night - he suggested that we instead block variant ser/de, however I took a second look today and, unless I'm missing something, I believe this route is preferable:

- This will fail earlier during planning rather than execution
- Blocking during ser/de may require more code changes, because pandas udfs don't seem to have the same codepath as non-pandas udfs.
- the `_parse_datatype_string` also was made to work with spark connect (which was @HyukjinKwon's initial concern) in [this](https://github.com/apache/spark/commit/fa8aa571ad18441622bb7e3ac66032ab9e7cbc0a) pr

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala (hunk 5 above)
Comment (richardc-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala]: I moved it to `VariantExpressionEvalUtils`

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/types.py (hunk 4 above)
Comment (allisonwang-db) [reply python/pyspark/sql/types.py:197]: Can we make sure the behaivor of `_parse_datatype_string` is the same as the original `fromDDL`? My concern is that this might introduce unintentional behavior change for a public API. 
What's the error message if we do  `fromDDL(a variant)` without this change?

Comment (richardc-db) [reply python/pyspark/sql/types.py:197]: yeah makes sense. The function comment mentions that this was added for Spark 4.0.0 (which won't be GA until 2025 IIUC?) so I thought the risk of a breaking change for a public API would be a bit lower.

The error message without changing `fromDDL` for something like "a int v variant" would be
```
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.UNSUPPORTED_UDF_OUTPUT_TYPE] Cannot resolve "<lambda>(id)" due to data type mismatch: UDFs do not support 'STRUCT<a: INT, v: VARIANT>' as an output data type. SQLSTATE: 42K09
```
which is pretty confusing for `fromDDL` in my opinion (i.e. why does `fromDDL` need to call a UDF in the first place?)

It seems that the UDF `returnType` implementation directly calls `_parse_datatype_string` [here](https://github.com/apache/spark/blob/master/python/pyspark/sql/udf.py#L370).

I added an additional test to test that the old udf behavior matches the new behavior

**Review by allisonwang-db — APPROVED** [review]
Thanks for adding the tests!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonUDFSuite.scala [hunk 6]
```diff
@@ -86,6 +86,61 @@ class PythonUDFSuite extends QueryTest with SharedSparkSession {
...
+    val testUdf = TestGroupedAggPandasUDFStringifiedMax(name = "pandas_udf")
+    checkError(
+      exception = intercept[AnalysisException] {
+        df.agg(testUdf(df("v"))).collect()
+      },
+      errorClass = "DATATYPE_MISMATCH.UNSUPPORTED_UDF_INPUT_TYPE",
+      parameters = Map("sqlExpr" -> "\"pandas_udf(v)\"", "dataType" -> "VARIANT"))
+  }
+
+  test("complex variant input to pandas grouped agg UDF") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonUDFSuite.scala:103]: To @richardc-db, @allisonwang-db , @HyukjinKwon .

This test case seems to miss `assume(shouldTestPandasUDFs)` mistakenly.

**Comment by dongjoon-hyun** [comment]
Here is a follow-up PR.
- https://github.com/apache/spark/pull/47599
